doc/SAIP4Extensions/SAI_P4EXT_API_Proposal.md

```diff
@@ -0,0 +1,279 @@
+
+/**
+ * @brief Set P4 Table entry attribute
+ *
+ * @param[in] p4ext_entry_id The P4 table id
+ * @param[in] attr Attribute
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_set_p4ext_entry_attribute_fn)(
+        _In_ sai_object_id_t p4ext_entry_id,
+        _In_ const sai_attribute_t *attr);
+
+/**
+ * @brief Get P4 entry attribute
+ *
+ * @param[in] p4ext_entry_id P4 table id
+ * @param[in] attr_count Number of attributes
+ * @param[inout] attr_list Array of attributes
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_get_p4ext_entry_attribute_fn)(
+        _In_ sai_object_id_t p4ext_entry_id,
+        _In_ uint32_t attr_count,
+        _Inout_ sai_attribute_t *attr_list);
+```
+
+### SAI P4 Extension API Table ###
+
+```cpp
+/**
+ * @brief P4Ext  methods table retrieved with sai_api_query()
+ **/
+typedef struct _sai_p4ext_api_t
+{
+    sai_create_p4ext_entry_fn                     create_p4ext_entry;
+    sai_remove_p4ext_entry_fn                     remove_p4ext_entry;
+    sai_set_p4ext_entry_attribute_fn              set_p4ext_entry_attribute;
+    sai_get_p4ext_entry_attribute_fn              get_p4ext_entry_attribute;
+} sai_p4ext_api_t;
+
+```
+
+# API Usage #
+
+Consider a scenario where a Vendor A wishes to extend the SAI pipeline functionality with a custom feature not yet supported in SAI on a target that is P4 compliant. Let say this new feature classifies incoming packets based on packet field tuples SIP,DIP and sets the traffic class tc. As mentioned earlier the proposal does not deal with the implementation details of data pipeline by each individual vendor, thus the relative position of this table in the SAI pipeline is not part of the specification and is vendor dependent. Its upto the vendor to decide whether this table will come after the SAI QoS tables or before and resolve any conflict that would arise from the actions of the QoS tables and the new P4 Extension Table. This simple feature can be expressed in P4 as below.
+
+```
+table flow_classification {
+    key = {
+        ig_md.lkp.ip_src_addr[95:64] : ternary @name("src_addr");
+        ig_md.lkp.ip_dst_addr[95:64] : ternary @name("dst_addr");
+    }
+    actions = {
+        set_tc;
+    }
+}
+```
+
+## Creatoing a P4 Extension Table Entry ##
+
+```cpp
+#define P4_TABLE_NAME "flow_classification"
+#define P4_TABLE_MATCH_FIELDS "{\"priority\":100,\"src_addr\":\"10.1.1.0&255.255.255.0\"}"
+#define P4_TABLE_ACTION "set_tc" 
+#define P4_TABLE_ACTION_PARAMETERS "{\"tc\":\"4\"}"
+
+sai_object_id_t p4ext_entry_id = {};
+uint8_t num_attrs = 4;
+sai_attribute_id_t p4ext_entry_attrs[num_attrs] = {};
+p4ext_entry_attrs[0].id = (sai_attr_id_t) SAI_P4EXT_ENTRY_ATTR_TABLE_ID;
+p4ext_entry_attrs[0].value.s8list.count = strlen(P4_TABLE_NAME);
+p4ext_entry_attrs[0].value.s8list.value = (sai_s8_list_t.list)P4_TABLE_NAME;  	//P4 Table Name
```

Comment:
The formatting looks weird here. Move the comments to the top to make it easier to read